src/network.rs

```diff
@@ -0,0 +1,508 @@
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Ok(records)).ok();
+							}
+						},
+						Err(err) => {
+							if let Some(QueryChannel::GetRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					QueryResult::PutRecord(result) => match result {
+						Ok(PutRecordOk { .. }) => {
+							if let Some(QueryChannel::PutRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Ok(())).ok();
+							}
+						},
+						Err(err) => {
+							if let Some(QueryChannel::PutRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					QueryResult::Bootstrap(result) => match result {
+						Ok(BootstrapOk { .. }) => {
+							if let Some(QueryChannel::Bootstrap(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Ok(())).ok();
+							}
+						},
+						Err(err) => {
+							if let Some(QueryChannel::Bootstrap(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					_ => {},
+				},
+				_ => {},
+			},
+			SwarmEvent::NewListenAddr { address, .. } => {
+				let local_peer_id = *self.swarm.local_peer_id();
+				info!(
+					"Local node is listening on {:?}",
+					address.with(Protocol::P2p(local_peer_id.into()))
+				);
+			},
+			SwarmEvent::ConnectionEstablished {
+				peer_id, endpoint, ..
+			} => {
+				if endpoint.is_dialer() {
+					if let Some(ch) = self.pending_dials.remove(&peer_id) {
+						let _ = ch.send(Ok(()));
+					}
+				}
+				// this event is of interest,
+				// pass event to output event stream
+				self.event_sender
+					.send(Event::ConnectionEstablished { peer_id, endpoint })
+					.await
+					.expect("Event receiver not to be dropped.");
+			},
+			SwarmEvent::OutgoingConnectionError { peer_id, error } => {
+				if let Some(peer_id) = peer_id {
+					if let Some(ch) = self.pending_dials.remove(&peer_id) {
+						_ = ch.send(Err(error.into()));
+					}
+				}
+			},
+			SwarmEvent::Dialing(peer_id) => info!("Dialing {}", peer_id),
+			_ => {},
+		}
+	}
+
+	async fn handle_command(&mut self, command: Command) {
+		match command {
+			Command::StartListening { addr, sender } => {
+				_ = match self.swarm.listen_on(addr) {
+					Ok(_) => sender.send(Ok(())),
+					Err(e) => sender.send(Err(e.into())),
+				}
+			},
+			Command::AddAddress {
+				peer_id,
+				peer_addr,
+				sender,
+			} => {
+				self.swarm
+					.behaviour_mut()
+					.kademlia
+					.add_address(&peer_id, peer_addr.clone());
+				self.pending_kad_routing.insert(peer_id, sender);
+			},
+			Command::Dial {
+				peer_id,
+				peer_addr,
+				sender,
+			} => {
+				// Check if peer is not already connected
+				// Dialing connected peers could happen during
```

Comment:
Was this tested with the actual data insertion or retrieval? I've seen plenty of (re)dials of the already connected peers, not sure how they're handled...

Reply:
These are dials that happen explicitly, on our command to the libp2p swarm. We can control these.
Those dials you've seen are implicit, coming from swarm itself, handling all network behavior with internal dialer schedulers

Reply:
These are commanded dials.
Not the ones coming from swarm.